lib/AppInfo/Application.php

```diff
@@ -21,110 +21,57 @@
 
 namespace OCA\Guests\AppInfo;
 
-use OC\Server;
-use OCA\Guests\Config;
+use OCA\Files\Event\LoadAdditionalScriptsEvent;
+use OCA\Guests\Listener\LoadAdditionalScriptsListener;
 use OCA\Guests\GroupBackend;
-use OCA\Guests\GuestManager;
 use OCA\Guests\Hooks;
 use OCA\Guests\Listener\ShareAutoAcceptListener;
 use OCA\Guests\Notifications\Notifier;
 use OCA\Guests\RestrictionManager;
 use OCA\Guests\UserBackend;
 use OCP\AppFramework\App;
-use OCP\EventDispatcher\IEventDispatcher;
+use OCP\AppFramework\Bootstrap\IBootstrap;
+use OCP\AppFramework\Bootstrap\IRegistrationContext;
+use OCP\AppFramework\Bootstrap\IBootContext;
+use OCP\AppFramework\IAppContainer;
+use OCP\IServerContainer;
 use OCP\IUser;
 use OCP\Notification\IManager as INotificationManager;
 use OCP\Share\Events\ShareCreatedEvent;
 
-class Application extends App {
+class Application extends App implements IBootstrap {
 	public const APP_ID = 'guests';
 
 	public function __construct(array $urlParams = []) {
 		parent::__construct(self::APP_ID, $urlParams);
 	}
 
-	public function setup() {
-		$this->setupGuestManagement();
-		$this->setupGuestRestrictions();
-		$this->setupNotifications();
-		$this->setupShareAccepting();
+	public function register(IRegistrationContext $context): void {
+		$context->registerEventListener(LoadAdditionalScriptsEvent::class, LoadAdditionalScriptsListener::class);
+		$context->registerEventListener(ShareCreatedEvent::class, ShareAutoAcceptListener::class);
 	}
 
-	public function lateSetup() {
-		$this->getRestrictionManager()->lateSetupRestrictions();
+	public function boot(IBootContext $context): void {
+		$this->setupGuestManagement($context->getAppContainer(), $context->getServerContainer());
```

Comment:
The user and group backend would fit the register method better to make sure that they are properly added to the manager classes already during the boot method of any app and also before the session is initialized.

Reply:
I've considered this but IRegistrationContext has no method for user and group backends, so I'd need to cheat there and call the container directly. If that's ok I can make the change.

Reply:
Can we do both? Like for now call the container impurely, but also add the missing API(s) for user and group back-end registration in `register`? The only possibly downside of adding that API now is that we still have a some technical dept in the user back-end API, so we might have to *break* that API once the overhaul is done. Ref https://help.nextcloud.com/t/user-backend-api-overhaul/94896.

Reply:
ok, done. I've cheated

Reply:
> Can we do both? Like for now call the container impurely, but also add the missing API(s) for user and group back-end registration in `register`? The only possibly downside of adding that API now is that we still have a some technical dept in the user back-end API, so we might have to _break_ that API once the overhaul is done. Ref https://help.nextcloud.com/t/user-backend-api-overhaul/94896.

While I agree that it makes sense, I didn't want to go down that rabbit hole as part of this PR, which fixes an important issue with initialization order.